File: Java/ComplexNumber.java
class ComplexNumber {

    private double real;
    private double imag;

    public ComplexNumber() {
        this.real = 0;
        this.imag = 0;
    }

    public ComplexNumber(double num1, double num2) {
        this.real = num1;
        this.imag = num2;
    }

    public ComplexNumber(ComplexNumber cn) {
        this.real = cn.real;
        this.imag = cn.imag;
    }

    public void setReal(double input) {
        this.real = input;
    }

    public void setImag(double input) {
        this.imag = input;
    }

    public double getReal() {
        return this.real;
    }

    public double getImag() {
        return this.imag;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof ComplexNumber) {
			ComplexNumber rnObj = (ComplexNumber) obj;
			if (rnObj.real == this.real &&
            rnObj.imag == this.imag) {
                return true;
            } else {
                return false;
            } 
		} else {
			return false;
		}
    }

    @Override
    public String toString() {
        boolean real_isNegative = false;
        boolean imag_isNegative = false;
        boolean real_isZero = false;
        boolean imag_isZero = false;
        String output = "";

        if (this.real < 0) {
           real_isNegative = true; 
        } else if (this.real == 0) {
            real_isZero = true;
        }

        if (this.imag < 0) {
            imag_isNegative = true;
        } else if (this.imag == 0) {
            imag_isZero = true;
        }

        double tempReal = Math.abs(this.real);
        double tempImag = Math.abs(this.imag);
        
        if (real_isZero && imag_isZero) {
            output = "0";
        } else if (!real_isZero && imag_isZero) {
            if (real_isNegative) {
                output = "-" + tempReal;
            } else {
                output = "" + tempReal;
            }
        } else if (real_isZero && !imag_isZero) {
            if (imag_isNegative) {
                output = "-" + tempImag + "i";
            } else {
                output = "" + tempImag + "i";
            }
        } else {
            if (real_isNegative && imag_isNegative) {
                output = "-" + tempReal + " - " + tempImag + "i";
            } else if (!real_isNegative && imag_isNegative) {
                output = "" + tempReal + " - " + tempImag + "i";
            } else if (real_isNegative && !imag_isNegative) {
                output = "-" + tempReal + " + " + tempImag + "i";
            } else {
                output = "" + tempReal + " + " + tempImag + "i";
            }
        }

        return output;
    }

    public ComplexNumber add(ComplexNumber input) {
        double tempReal = this.real + input.real;
        double tempImag = this.imag + input.imag;
        ComplexNumber output = new ComplexNumber(tempReal, tempImag);
        return output;
    }

    public ComplexNumber sub(ComplexNumber input) {
        double tempReal = this.real - input.real;
        double tempImag = this.imag - input.imag;
        ComplexNumber output = new ComplexNumber(tempReal, tempImag);
        return output;
    }

    public ComplexNumber mult(ComplexNumber input) {
        double tempReal = this.real * input.real;
        double tempImag = this.imag * input.imag;
        ComplexNumber output = new ComplexNumber(tempReal, tempImag);
        return output;
    }

    public ComplexNumber div(ComplexNumber input) throws ArithmeticException {
        
        try {
            double tempReal = this.real / input.real;
            double tempImag = this.imag / input.imag;
            ComplexNumber output = new ComplexNumber(tempReal, tempImag);
            return output;
        } catch(ArithmeticException e) {
            throw new ArithmeticException("Cannot divide by 0");
        }
    }

    public double mag() {
        double magnitude = Math.sqrt(this.real*this.real + this.imag*this.imag);
        return magnitude;
    }

    public ComplexNumber conj() {
        double tempImag = this.imag * -1;
        ComplexNumber output = new ComplexNumber(this.real, tempImag);
        return output;
    }

    public ComplexNumber sqrt() {
        boolean real_isNegative = false;
        boolean imag_isNegative = false;
        double tempReal = 0;
        double tempImag = 0;
        if (this.real < 0) real_isNegative = true;
        if (this.imag < 0) real_isNegative = true;
        
        // -- Check for logic of sqrt(imag)
        if (!real_isNegative && !imag_isNegative) {
            tempReal = Math.sqrt(Math.abs(this.real));
            tempImag = Math.sqrt(Math.abs(this.imag));
        } else if (!real_isNegative && imag_isNegative) {
            tempReal = Math.sqrt(Math.abs(this.real))
                - Math.sqrt(Math.abs(this.imag));
            tempImag = 0;
        } else if (real_isNegative && !imag_isNegative) {
            tempReal = 0;
            tempImag = Math.sqrt(Math.abs(this.imag)) 
                + Math.sqrt(Math.abs(this.real));
        } else {
            tempReal = -1 * Math.sqrt(Math.abs(this.imag));
            tempImag = Math.sqrt(Math.abs(this.real));
        }
        
        ComplexNumber output = new ComplexNumber(tempReal, tempImag);
        return output;
    }
}

File: C++/ComplexNumber.cpp
#include "ComplexNumber.h"

#include <string>
#include <typeinfo>
#include <cmath>

using namespace std;

double real;
double imag;

ComplexNumber::ComplexNumber(): real(0), imag(0) {}

ComplexNumber::ComplexNumber(double realNum, double imagNum) {
    this->real = realNum;
    this->imag = imagNum;
}

ComplexNumber::ComplexNumber(const ComplexNumber &src) {
    this->real = src.real;
    this->imag = src.imag;
}

ComplexNumber::~ComplexNumber() {}

void ComplexNumber::setReal(double num) {
    this->real = num;
}

void ComplexNumber::setImag(double num) {
    this->imag = num;
}

double ComplexNumber::getReal() {
    return this->real;
}

double ComplexNumber::getImag() {
    return this->imag;
}

string ComplexNumber::toString() {
    bool realNegative = false;
    bool imagNegative = false;
    if (this->real < 0) realNegative = true;
    if (this->imag < 0) imagNegative = true;
    
    double realOutput = std::abs(this->real);
    double imagOutput = std::abs(this->imag);

    if (realOutput == 0.0 && imagOutput == 0.0) {
        return "0";
    } else if (realOutput != 0.0 && imagOutput == 0.0) {
        if (realNegative) 
            return "-" + to_string(realOutput);
        return to_string(realOutput);
    } else if (realOutput == 0.0 && imagOutput != 0.0) {
        if (imagNegative)
            return "0 - " + to_string(imagOutput) + "i";
        return to_string(imagOutput) + "i";
    }
    if (realNegative && imagNegative) {
        return "-" + to_string(realOutput) + " - " + to_string(imagOutput) + "i";
    } else if (realNegative) {
        return "-" + to_string(realOutput) + " + " + to_string(imagOutput) + "i";
    } else if (imagNegative) {
        return to_string(realOutput) + " - " + to_string(imagOutput) + "i";
    } else {
        return to_string(realOutput) + " + " + to_string(imagOutput) + "i";
    }
}

ComplexNumber ComplexNumber::add(ComplexNumber rhs) {
    double realOutput = this->real + rhs.real;
    double imagOutput = this->imag + rhs.imag;
    ComplexNumber output(realOutput, imagOutput);
    return output;
}

ComplexNumber ComplexNumber::sub(ComplexNumber rhs) {
    double realOutput = this->real - rhs.real;
    double imagOutput = this->imag - rhs.imag;
    ComplexNumber output(realOutput, imagOutput);
    return output;
}

ComplexNumber ComplexNumber::mult(ComplexNumber rhs) {
    double realOutput = this->real * rhs.real;
    double imagOutput = this->imag * rhs.imag;
    ComplexNumber output(realOutput, imagOutput);
    return output;
}

ComplexNumber ComplexNumber::div(ComplexNumber rhs) {
    char divZeroMessage[] = "Cannot be divisible by 0";
    char *divZeroException = divZeroMessage;
    double realOutput = 0;
    double imagOutput = 0;
    try {
        if (rhs.real == 0 || rhs.imag == 0) {
            throw divZeroException;
            exit;
        } else {
            realOutput = this->real / rhs.real;
            imagOutput = this->imag / rhs.imag;
            ComplexNumber output(realOutput, imagOutput);
            return output;
        }
    } catch (const char* exception) {
        cout << "Exception: " << exception << endl;
    }
} 

double ComplexNumber::mag() {
    double realOutput = this->real;
    double imagOutput = this->imag;
    double output = std::sqrt(std::pow(realOutput, 2) + std::pow(imagOutput, 2));
    return output;
}

ComplexNumber ComplexNumber::conj() {
    double imagOutput = 0 - this->imag;
    ComplexNumber output(this->real, imagOutput);
}

ComplexNumber ComplexNumber::sqrt() {
    // -- sqrt(-1) = i
    // -- sqrt(i) = -1
    if (this->real < 0 && this->imag < 0) {
        double realOutput = std::sqrt(std::abs(this->imag));
        double imagOutput = std::sqrt(std::abs(this->real));
        ComplexNumber output(realOutput, imagOutput);
        return output;
    } else if (this->real < 0) {
        double realOutput = 0;
        double imagOutput = std::sqrt(abs(this->real)) + std::sqrt(this->imag);
        ComplexNumber output(realOutput, imagOutput);
        return output;
    } else if (this->imag < 0) {
        double realOutput = std::sqrt(this->real) + std::sqrt(std::abs(this->imag));
        double imagOutput = 0;
        ComplexNumber output(realOutput, imagOutput);
        return output;
    }
    double realOutput = std::sqrt(this->real);
    double imagOutput = std::sqrt(this->imag);
    ComplexNumber output(realOutput, imagOutput);
    return output;
}

bool ComplexNumber::equals(ComplexNumber& rhs) {
    if (typeid(*this) != typeid(rhs))
        return false;
    if (this->real == rhs.real && this->imag == rhs.imag)
        return true;
    return false;
}